### DISCUSSION TITLE
widget with data-dependent compose method

### DISCUSSION THREAD

**Discussion by dbieber** [Textualize/textual#3968] [Q&A]
Still wrapping my head around the textual way of doing things. In this simple example, the MultiLabelWidget's compose method is dependent on the reactive num_labels. But compose is not called when num_labels changes. I'd like, when a click occurs, for a fourth label to be added. What is the textual way of having a Widget who's composition depends on its reactive data?

```python
from textual.app import App
from textual.app import ComposeResult
from textual.reactive import reactive
from textual.widget import Widget
from textual.widgets import Label


class MultiLabelWidget(Widget):

  num_labels = reactive(3)

  def compose(self) -> ComposeResult:
    for i in range(self.num_labels):
      yield Label(f"Label {i}", id=f"label_{i}")


class ExampleApp(App):

  def compose(self) -> ComposeResult:
    yield MultiLabelWidget()

  def on_click(self):
    self.query_one("#label_0").update("Yes, clicks are working.")
    self.query_one(MultiLabelWidget).num_labels += 1


if __name__ == "__main__":
  app = ExampleApp()
  app.run()
```

<img width="794" alt="image" src="https://github.com/Textualize/textual/assets/892765/72df855a-d31d-41a1-b777-4c5014d111c4">

---

I know that I can achieve this in this simple example with mount, but iiuc there should be a declarative approach as well, and that might be preferable over managing mount/unmount calls myself (?), which seems like it could get quite complex quickly.

**Comment by dbieber** [comment]
`reactive(3, layout=True)` does not change the app's behavior

**Reply by dbieber** [reply]
Closer to my actual use case: I'd like a Widget that displays a list of strings. It should display one widget if the list is empty, and a collection of different widgets (one per string) if the list is non-empty. And it should handle this list changing.

**Reply by dbieber** [reply]
Maybe what I need is:
* a ContentSwitcher to switch between the no-item and 1+ item cases
* a ListView for the 1+ item case

Going to learn about ContentSwitcher next.

**Reply by dbieber** [reply]
For my own notes... this first example runs, but it's not actually reactive to self.strings getting set.

```python
from textual.app import App, ComposeResult
from textual.reactive import reactive
from textual.widgets import Label, ListView, ListItem, ContentSwitcher


class ListDisplayApp(App):

    strings = reactive(lambda: ["Initial item"])

    def compose(self) -> ComposeResult:
        with ContentSwitcher(id="list_switcher", initial="list_view") as switcher:
            yield Label("The list is empty.", id="empty_label")
            yield ListView(*(ListItem(Label(string)) for string in self.strings), id="list_view")

    def on_mount(self) -> None:
        self.set_timer(1, callback=self.add_string)

    def add_string(self):
        self.strings.append(f"Item {len(self.strings) + 1}")
        switcher = self.query_one(ContentSwitcher)
        switcher.current = "list_view" if self.strings and len(self.strings) < 5 else "empty_label"
        list_view = self.query_one(ListView)
        list_view.clear()
        list_view.extend([ListItem(Label(string)) for string in self.strings])
        self.set_timer(1, callback=self.add_string)

if __name__ == "__main__":
    app = ListDisplayApp()
    app.run()
```

whereas with this small modification, I encounter the error `NoMatches: No nodes match <DOMQuery query='ContentSwitcher'>`:

```python
from textual.app import App, ComposeResult
from textual.reactive import reactive
from textual.widgets import Label, ListView, ListItem, ContentSwitcher


class ListDisplayApp(App):

    strings = reactive(lambda: ["Initial item"])

    def compose(self) -> ComposeResult:
        with ContentSwitcher(id="list_switcher", initial="list_view") as switcher:
            yield Label("The list is empty.", id="empty_label")
            yield ListView(*(ListItem(Label(string)) for string in self.strings), id="list_view")

    def on_mount(self) -> None:
        self.set_timer(1, callback=self.add_string)

    def add_string(self):
        self.strings.append(f"Item {len(self.strings) + 1}")

    def watch_strings(self):  # *** Modification here ***
        switcher = self.query_one(ContentSwitcher)
        switcher.current = "list_view" if self.strings and len(self.strings) < 5 else "empty_label"
        list_view = self.query_one(ListView)
        list_view.clear()
        list_view.extend([ListItem(Label(string)) for string in self.strings])
        self.set_timer(1, callback=self.add_string)

if __name__ == "__main__":
    app = ListDisplayApp()
    app.run()
```

**Reply by dbieber** [reply]
Seems like my issue in that second example is I should not be accessing self.strings to read the value of a reactive(?). Instead I should be using watch_strings and updating the widget using the value there.

So for example this runs without error:
```python
from textual.app import App, ComposeResult
from textual.containers import Container
from textual.reactive import reactive
from textual.widget import Widget
from textual.widgets import Label, ListView, ListItem, ContentSwitcher


class ListWidget(Widget):
    num_strings = reactive(4)

    def compose(self) -> ComposeResult:
        yield ListView()

    def watch_num_strings(self, num_strings_value):
        list_view = self.query_one(ListView)
        list_view.clear()
        children = [ListItem(Label(f"Label {i}")) for i in range(num_strings_value)]
        list_view.extend(children)


class ListDisplayApp(App):

    def compose(self) -> ComposeResult:
        yield ListWidget(id="list")

if __name__ == "__main__":
    app = ListDisplayApp()
    app.run()
```

whereas this does not:
```python
from textual.app import App, ComposeResult
from textual.containers import Container
from textual.reactive import reactive
from textual.widget import Widget
from textual.widgets import Label, ListView, ListItem, ContentSwitcher


class ListWidget(Widget):
    num_strings = reactive(4)

    def compose(self) -> ComposeResult:
        children = [ListItem(Label(f"Label {i}")) for i in range(self.num_strings)]
        yield ListView(*children)

    def watch_num_strings(self, num_strings_value):
        label = self.query_one(ListView)


class ListDisplayApp(App):

    def compose(self) -> ComposeResult:
        yield ListWidget(id="list")

if __name__ == "__main__":
    app = ListDisplayApp()
    app.run()
```
My theory is that accessing self.num_strings from within compose causes watch_num_strings to get called before compose has finished, thereby invoking the query before the DOM is ready.
The error I get is `NoMatches: No nodes match <DOMQuery query='ListView'>`

-------

Now that I understand as much, I can fix my initial example:

```python
from textual.app import App, ComposeResult
from textual.reactive import reactive
from textual.widgets import Label, ListView, ListItem, ContentSwitcher


class ListDisplayApp(App):
    strings = reactive(["Initial item"])

    def compose(self) -> ComposeResult:
        with ContentSwitcher(id="list_switcher", initial="list_view") as switcher:
            yield Label("The list is empty.", id="empty_label")
            yield ListView(id="list_view")

    def add_string(self):
        self.strings = self.strings + [f"Item {len(self.strings) + 1}"]

    def watch_strings(self, strings_value):
        switcher = self.query_one(ContentSwitcher)
        switcher.current = "list_view" if self.strings and len(strings_value) < 5 else "empty_label"
        list_view = self.query_one(ListView)
        list_view.clear()
        list_view.extend([ListItem(Label(string)) for string in strings_value])
        self.set_timer(1, callback=self.add_string)

if __name__ == "__main__":
    app = ListDisplayApp()
    app.run()
```

**Reply by davep** [reply]
Your question seems to have evolved a fair bit from when first asked; is there anything that is needing an answer now?

**Reply by dbieber** [reply]
My remaining open questions are:

1. **Does a widget's compose ever get called again after the initial call to compose, allowing you to change the composition of a Widget in response to changes to data?** The [docs](https://textual.textualize.io/how-to/render-and-compose/) make it sound like render and compose are almost interchangeable, with the difference being that compose returns other Widgets whereas render returns a renderable or string. However, in practice it seems like Widgets get rerendered when data changes, but not recomposed. Is that a textual bug, a docs issue, or just my misunderstanding?

2. Is it a textual bug that the following code snippet errors? The error this gives is `NoMatches: No nodes match <DOMQuery query='ListView'>`.
```python
from textual.app import App, ComposeResult
from textual.containers import Container
from textual.reactive import reactive
from textual.widget import Widget
from textual.widgets import Label, ListView, ListItem, ContentSwitcher


class ListWidget(Widget):
    num_strings = reactive(4)

    def compose(self) -> ComposeResult:
        children = [ListItem(Label(f"Label {i}")) for i in range(self.num_strings)]
        yield ListView(*children)

    def watch_num_strings(self, num_strings_value):
        label = self.query_one(ListView)


class ListDisplayApp(App):

    def compose(self) -> ComposeResult:
        yield ListWidget(id="list")

if __name__ == "__main__":
    app = ListDisplayApp()
    app.run()
```
IIUC the issue in this code snippet is that accessing self.num_strings causes watch_num_strings to be called before the DOM is ready, so the query_one line fails. However, self.num_strings doesn't modify num_strings, so maybe watch_num_strings shouldn't be called yet.

3. Is there a more declarative / idiomatic approach to making a widget that updates when a list changes? The approach I found that works involves having that widget explicitly remove and add DOM elements when the list changes. This makes things declarative in its parents (which can simply set the value of the list), but imperative in its own handling of the changing data. I'm wondering if there's a way to replace this imperative bit with something declarative (i.e. if compose were called again, this would do the trick).

**Reply by davep** [reply]
Thanks, that makes it a wee bit easier to target what was left needing answering. :-)

1. `compose` is only called once, not multiple times, and especially not in response to a reactive changing. If you want to make changes to the DOM of a widget I'd suggest a revisit to the [latter part of the tutorial where it introduces you to dynamic widgets](https://textual.textualize.io/tutorial/#dynamic-widgets). If you've not read it yet you may also find [the howto on`render` and `compose`](https://textual.textualize.io/how-to/render-and-compose/) helpful in explaining the key differences.
2. No, the code is working as you'd expect; but I suspect there's a subtle issue you won't have picked up on yet. Unless you tell a reactive to *not* fire its watch methods when you first assign to it, the watch will fire when it's first assigned. In your case `watch_num_strings` will be called before the DOM has been mounted, and you're attempting to query the DOM from your watch. So here you'll want to change your code to initialise `num_strings` to `reactive(4, init=False)`, or you could change your `watch_num_strings` method so that it checks [if the widget has been mounted yet](https://textual.textualize.io/api/widget/#textual.widget.Widget.is_mounted).
3. Hopefully the answers to 1 and 2 will help with this.

**Reply by dbieber** [reply]
Your answers to 1 and 3 are helpful and make sense (though they are a bit disappointing too, and I'm not sure I understand the reasoning behind this design decision yet). 👍 

The behavior in 2 still confuses me. Indeed, if I set init=False, then watch_num_strings isn't called until after the DOM has been mounted, which is consistent with what you're saying. However, if I instead leave init=True but remove the reference to self.num_strings in compose, this _also_ prevents watch_num_strings from being called before the DOM has been mounted, and this is confusing.

Code illustrating this:
```python
from textual.app import App, ComposeResult
from textual.containers import Container
from textual.reactive import reactive
from textual.widget import Widget
from textual.widgets import Label, ListView, ListItem, ContentSwitcher


class ListWidget(Widget):
    num_strings = reactive(4, init=True)

    def compose(self) -> ComposeResult:
        children = []
        children = [ListItem(Label(f"Label {i}")) for i in range(4)]
        yield ListView(*children)

    def watch_num_strings(self, num_strings_value):
        label = self.query_one(ListView)


class ListDisplayApp(App):

    def compose(self) -> ComposeResult:
        yield ListWidget(id="list")

if __name__ == "__main__":
    app = ListDisplayApp()
    app.run()
```

**Reply by davep** [reply]
I'm not very familiar with the implementation details of reactives, but my suspicion here would be that the initial watch event on assignment is fired the first time the reactive is referenced; having removed any reference you've likely removed any situation in which it gets initialised; and so no call to `watch_num_strings`.

**Reply by dbieber** [reply]
Thanks for taking the time to answer my questions.

Regarding 1 + 3, I am contemplating creating a Widget base class or mixin that calls compose when a reactive changes, and then removes and mounts sub-widgets accordingly in order to make the DOM match whatever compose returns. (Still in the contemplation phase; I am simply sharing this here in case it sparks discussion or if there's a reason not to do this.)

**Reply by dbieber** [reply]
It seems the simplest (though perhaps not most efficient) way to recompose a Widget is:
```python
    self.query(None).remove()
    for item in self.compose():
      self.mount(item)
```

(Edit: Ah, I see this doesn't actually always work)
(Edit 2: `for item in _compose.compose(self)` does the trick 👍 )
(Edit 3: But as might be expected, it causes the screen to flash as everything is removed, before everything is readded, which isn't great.)
(Edit 4: The self.app.batch_update() context manager looks like it should resolve this, but unfortunately doesn't seem to.)

New Q:
**Is there a way to have a watch_all method that triggers whenever _any_ reactive is updated?**

Reason for this Q:
I'm considering trying out having my widgets recompose whenever any of their reactives change, to give a more declarative programming paradigm.

(Also lmk if you'd prefer I open new discussions for new Qs rather than continuing the discussion.)

**Reply by dbieber** [reply]
After all the edits in my previous message, here's what I'm left with:
```python
  def watch_num_strings(self) -> None:
    with self.app.batch_update():
      self.query(None).remove()
      self.mount_all(_compose.compose(self))
```
It does flicker, which is unexpected because batch_update seems like it ought to avoid the flicker.

**Reply by davep** [reply]
In the code you post there, I'd expect flicker, as you're not awaiting the removals or the mounts. Consider this:

```python
from textual.app import App, ComposeResult
from textual.widgets import Label, Input

class RemountApp(App[None]):

    BINDINGS = [
        ("f1", "sync"),
        ("f2", "async")
    ]

    def compose(self) -> ComposeResult:
        for n in range(1_000):
            yield Label(f"This is label {n}")
            yield Input(placeholder=f"This is input {n}")

    def action_sync(self) -> None:
        with self.batch_update():
            self.query("Label, Input").remove()
            self.mount_all(self.compose())

    async def action_async(self) -> None:
        with self.batch_update():
            await self.query("Label, Input").remove()
            await self.mount_all(self.compose())

if __name__ == "__main__":
    RemountApp().run()
```

**Reply by dbieber** [reply]
Yes, that avoids the flicker! Thank you.

**Reply by dbieber** [reply]
I'm next working on a different sort of flicker: I have a `self.call_after_refresh(container.scroll_end, animate=False)` after the batch_update() call, which works, but it first goes to the top on each update before doing the scroll to the bottom.
